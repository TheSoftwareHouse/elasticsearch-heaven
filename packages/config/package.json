{
  "name": "@elasticsearch-heaven/config",
  "version": "0.0.1"
}
