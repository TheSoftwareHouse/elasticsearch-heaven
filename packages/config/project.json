{
  "root": "packages/config",
  "sourceRoot": "packages/config/src",
  "projectType": "library",
  "targets": {
    "build": {
      "executor": "@nrwl/node:package",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/packages/config",
        "tsConfig": "packages/config/tsconfig.lib.json",
        "packageJson": "packages/config/package.json",
        "main": "packages/config/src/index.ts",
        "assets": ["packages/config/*.md"]
      }
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["packages/config/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/packages/config"],
      "options": {
        "jestConfig": "packages/config/jest.config.js",
        "passWithNoTests": true
      }
    }
  },
  "tags": []
}
