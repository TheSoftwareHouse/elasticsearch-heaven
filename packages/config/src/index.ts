export * from './resolveConfig';
export * from './types';
export * from './readConfig';
