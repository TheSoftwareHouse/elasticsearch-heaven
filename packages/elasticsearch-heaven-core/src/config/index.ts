export * from './resolveConfig';
export * from './readConfig';
