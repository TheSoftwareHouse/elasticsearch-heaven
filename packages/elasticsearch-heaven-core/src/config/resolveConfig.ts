import {
  Config,
  Logger,
  MigrationOptions,
  MigrationsLockStorage,
  MigrationsStorage,
} from '@tshio/elasticsearch-heaven-types';
import { EsMigrationLock, EsMigrationStorage } from '../migrations';
import { createConsoleLogger } from '../logger';
import { Client } from '@elastic/elasticsearch';

export interface ResolvedConfig extends Pick<Config, 'debug'> {
  logger: Logger;
  client: Client;
  migrations: MigrationOptions & {
    migrationsStorage: MigrationsStorage;
    lock: MigrationsLockStorage;
  };
}

export const resolveConfig = (config: Config): ResolvedConfig => {
  const logger = config.logger ?? createConsoleLogger(config.debug);

  const client =
    'client' in config ? config.client : new Client(config.clientOptions);

  return {
    ...config,
    logger,
    client,
    migrations: resolveMigrations(config.migrations, client, logger),
  };
};

const resolveMigrations = (
  config: MigrationOptions,
  client: Client,
  logger: Logger
): ResolvedConfig['migrations'] => {
  return {
    ...config,
    migrationsStorage:
      'migrationsStorage' in config
        ? config.migrationsStorage
        : new EsMigrationStorage(
            config.migrationsIndexName ?? 'migrations',
            client,
            logger
          ),
    lock:
      'lock' in config
        ? config.lock
        : new EsMigrationLock(
            config.lockIndexName ?? 'migrations_lock',
            client,
            logger
          ),
  };
};
