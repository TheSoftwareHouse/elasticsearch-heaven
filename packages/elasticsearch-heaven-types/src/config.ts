import { MigrationOptions } from './migrations';
import { Client, ClientOptions } from '@elastic/elasticsearch';
import { Logger } from './logger';

interface BaseConfig {
  migrations: MigrationOptions;
  debug?: boolean;
  logger?: Logger;
}

export type Config =
  | (BaseConfig & {
      clientOptions: ClientOptions;
    })
  | (BaseConfig & { client: Client });
