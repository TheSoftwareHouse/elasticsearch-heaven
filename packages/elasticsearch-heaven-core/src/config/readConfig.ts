import { resolveConfig } from './resolveConfig';

const fileNames = ['tshio.config.js', 'tshio.config.ts'];

export const readConfig = async (configPath?: string) => {
  if (configPath) {
    const config = await import(configPath);

    return resolveConfig(config.default);
  }

  for (const fileName of fileNames) {
    try {
      const config = await import(`./${fileName}`);

      return resolveConfig(config.default);
    } catch (e) {
      // do nothing
    }
  }

  throw new Error('No config file found.');
};
