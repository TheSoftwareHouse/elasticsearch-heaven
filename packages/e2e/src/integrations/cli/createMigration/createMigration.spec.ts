import { $ } from 'zx';
import * as path from 'path';
import * as fs from 'fs-extra';
import config from './testConfig';

const configDir = path.resolve(__dirname, 'testConfig.ts');
const migrationsDir = config.migrations.migrationsDir;

const cleanup = () => {
  if (fs.pathExistsSync(migrationsDir)) {
    fs.emptyDirSync(migrationsDir);
    fs.rmdirSync(migrationsDir);
  }
};

describe('Create migration command', () => {
  beforeEach(() => {
    cleanup();
  });

  afterEach(() => {
    cleanup();
  });

  it('should create new migration file', async () => {
    await $`npx nx run elasticsearch-heaven-core:cli-create-migration --name="TestMigration" --config-path=${configDir}`;

    const files = fs.readdirSync(migrationsDir);

    expect(files).toHaveLength(1);

    const migration = fs
      .readFileSync(path.join(migrationsDir, files[0]))
      .toString();

    expect(migration).toMatchSnapshot();
  });
});
