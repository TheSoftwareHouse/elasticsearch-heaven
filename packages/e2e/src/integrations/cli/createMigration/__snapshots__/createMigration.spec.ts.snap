// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create migration command should create new migration file 1`] = `
"import {MigrationBag} from '@elasticsearch-heaven/migrations'

export async function up(bag: MigrationBag) {}

export async function down(bag: MigrationBag) {}"
`;
